From: <devin@debian.org>
Subject: Patch for integer underflow in RTSP streaming code, taken from upstream 
X-Upstream: commit 29447

diff --git a/stream/realrtsp/real.c b/stream/realrtsp/real.c
index 6539930..2e550ac 100644
--- a/stream/realrtsp/real.c
+++ b/stream/realrtsp/real.c
@@ -388,6 +388,8 @@ int real_get_rdt_chunk(rtsp_t *rtsp_session, char **buffer, int rdt_rawdata) {
     return (n <= 0) ? 0 : n;
   }
   rmff_dump_pheader(&ph, *buffer);
+  if (size < 12)
+    return 0;
   size-=12;
   n=rtsp_read_data(rtsp_session, (*buffer)+12, size);
 
