From: uau
Subject: stream/realrtsp/real.c: Fix another integer overflow
X-Upstream: commit r29455

--- trunk/stream/realrtsp/real.c	(revision 29454)
+++ trunk/stream/realrtsp/real.c	(revision 29455)
@@ -382,6 +382,8 @@
     ph.flags=0;
   *buffer = xbuffer_ensure_size(*buffer, 12+size);
   if(rdt_rawdata) {
+      if (size < 12)
+          return 0;
     n=rtsp_read_data(rtsp_session, *buffer, size-12);
     return (n <= 0) ? 0 : n;
   }
