From: Reimar
Subject: 100l, add forgotten BGR15 format to fmt-conversion.c table
Origin: upstream, r29117, http://lists.mplayerhq.hu/pipermail/mplayer-cvslog/2009-April/036930.html

--- a/fmt-conversion.c
+++ b/fmt-conversion.c
@@ -28,6 +28,7 @@ static const struct {
     {IMGFMT_BGR32, PIX_FMT_RGB32},
     {IMGFMT_BGR24, PIX_FMT_BGR24},
     {IMGFMT_BGR16, PIX_FMT_RGB565},
+    {IMGFMT_BGR15, PIX_FMT_RGB555},
     {IMGFMT_BGR8,  PIX_FMT_RGB8},
     {IMGFMT_BGR4,  PIX_FMT_RGB4},
     {IMGFMT_BGR1,  PIX_FMT_MONOBLACK},
