Author: Reinhard Tartler <siretart@tauware.de>
Subject: load_fontconfig by default
Status: Patch taken from upstream.

diff --git a/libvo/font_load_ft.c b/libvo/font_load_ft.c
index 20d673a..1cb00b1 100644
--- a/libvo/font_load_ft.c
+++ b/libvo/font_load_ft.c
@@ -66,17 +66,17 @@ float subtitle_font_thickness = 2.0;
 // 3 = diagonal
 int subtitle_autoscale = 3;
 
 int vo_image_width = 0;
 int vo_image_height = 0;
 int force_load_font;
 
 int using_freetype = 0;
-int font_fontconfig = 0;
+int font_fontconfig = 1;
 
 //// constants
 static unsigned int const colors = 256;
 static unsigned int const maxcolor = 255;
 static unsigned const	base = 256;
 static unsigned const first_char = 33;
 #define MAX_CHARSET_SIZE 60000
 

