Author: Reinhard Tartler <siretart@tauware.de>
Subject: allow compilation with newer x264

diff --git a/libmpcodecs/ve_x264.c b/libmpcodecs/ve_x264.c
index 04d0bd3..c4287af 100644
--- a/libmpcodecs/ve_x264.c
+++ b/libmpcodecs/ve_x264.c
@@ -207,7 +207,11 @@ static int config(struct vf_instance_s* vf, int width, int height, int d_width,
         mod->mux->bih->biSize= sizeof(BITMAPINFOHEADER) + extradata_size;
     }
     
+#if X264_BUILD >= 78
+    if (param.i_bframe > 1 && param.i_bframe_pyramid)
+#else
     if (param.i_bframe > 1 && param.b_bframe_pyramid)
+#endif
         mod->mux->decoder_delay = 2;
     else
         mod->mux->decoder_delay = param.i_bframe ? 1 : 0;

