# patch by Simon McVittie
--- mplayer-1.0pre6a.orig/libvo/vo_x11.c
+++ mplayer-1.0pre6a/libvo/vo_x11.c
@@ -39,7 +39,6 @@
 #include "sub.h"

 #include "postproc/swscale.h"
-#include "postproc/swscale_internal.h"       //FIXME
 #include "postproc/rgb2rgb.h"
 #include "libmpcodecs/vf_scale.h"

@@ -50,6 +49,10 @@
 #include "mplayer.h"
 #endif

+/* on PowerPC this includes altivec.h, which causes problems for
+sys/uio.h included by Gui/interface.h, if this one appears first */
+#include "postproc/swscale_internal.h"       //FIXME
+
 static vo_info_t info = {
     "X11 ( XImage/Shm )",
     "x11",
