Subject: Package desktop features with non-gui mplayer
 Non-gui mplayer is not currently packaged.
Author: Robbie Harwood (frozencemetery) <rharwood@club.cc.cmu.edu>
Forwarded: not-needed
--- a/etc/mplayer.desktop
+++ b/etc/mplayer.desktop
@@ -25,8 +25,9 @@
 Keywords=Player;Capture;Audio;Music;Song;Video;Movie;Film;Clip;DVD;VCD;CD;Disc;DVB;TV;
 Keywords[de]=Player;Wiedergabe;Capture;Mitschnitt;Audio;Musik;Lied;Video;Film;Clip;DVD;VCD;CD;Disc;Disk;DVB;TV;
 Icon=mplayer
-TryExec=gmplayer
-Exec=gmplayer %F
-Terminal=false
+TryExec=mplayer
+Exec=mplayer %U
+NoDisplay=true
+Terminal=true
 Categories=GTK;AudioVideo;Audio;Video;Player;TV;
 MimeType=application/mxf;application/x-netshow-channel;application/ogg;application/ram;application/vnd.adobe.flash.movie;application/vnd.rn-realmedia;application/x-shockwave-flash;application/smil;audio/aac;audio/ac3;audio/x-adpcm;audio/x-aiff;audio/AMR;audio/AMR-WB;audio/x-ape;audio/basic;audio/flac;audio/x-flac;audio/x-flac+ogg;audio/x-m4b;audio/x-matroska;audio/mp2;audio/mp4;audio/mpeg;audio/x-mpegurl;audio/x-ms-asx;audio/x-ms-wma;audio/x-musepack;audio/ogg;audio/vnd.rn-realaudio;audio/x-scpls;audio/x-voc;audio/x-vorbis+ogg;audio/x-wav;audio/x-wavpack;video/3gpp;video/3gpp2;video/dv;video/x-flic;video/x-flv;video/x-matroska;video/mp2t;video/mp4;video/mpeg;video/vnd.mpegurl;video/x-ms-asf;video/x-ms-wmv;video/x-msvideo;video/x-nsv;video/ogg;video/x-ogm+ogg;video/quicktime;video/vnd.rn-realvideo;video/x-theora+ogg;video/webm;x-content/audio-cdda;x-content/audio-dvd;x-content/video-dvd;x-content/video-svcd;x-content/video-vcd;
