--- libavcodec/Makefile.orig	2006-12-14 01:46:58.000000000 +0100
+++ libavcodec/Makefile	2006-12-14 01:54:17.000000000 +0100
@@ -301,6 +301,12 @@
 OBJS-$(CONFIG_H261_PARSER)             += h261.o
 OBJS-$(CONFIG_H263_PARSER)             += h263dec.o
 OBJS-$(CONFIG_H264_PARSER)             += h264.o
+#if --enable-debug is passed to configure, then
+#  -ffast-math -fomit-frame-pointer are omitted, 
+# but then compilation fails with error
+#cabac.h:513: error: can't find a register in class 'GENERAL_REGS' while reloading 'asm'
+#Googling around, it seems to be a widespread problem 
+h264.o : CFLAGS += -ffast-math -fomit-frame-pointer
 OBJS-$(CONFIG_MJPEG_PARSER)            += mjpeg.o
 OBJS-$(CONFIG_MPEG4VIDEO_PARSER)       += parser.o
 OBJS-$(CONFIG_MPEGAUDIO_PARSER)        += parser.o
