--- configure.orig	2006-10-12 10:37:56.000000000 +0200
+++ configure	2006-10-13 12:25:38.000000000 +0200
@@ -3748,16 +3748,12 @@
 echocheck "DirectFB"
 if test "$_directfb" = auto ; then
   _directfb=no
-  if linux && test -c /dev/fb0; then
+  if pkg-config --exists 'directfb' ; then
     cat > $TMPC <<EOF
 #include <directfb.h>
 int main(void) { IDirectFB *foo; DirectFBInit(0,0); return 0; }
 EOF
-  for _inc_tmp in "" -I/usr/local/include/directfb \
-      -I/usr/include/directfb -I/usr/local/include -I/usr/include; do
-    cc_check $_inc_tmp -ldirectfb && _directfb=yes && \
-      _inc_extra="$_inc_extra $_inc_tmp" && break
-  done
+    cc_check `pkg-config --libs --cflags directfb` && _directfb=yes
   fi
 fi
 
@@ -3766,29 +3762,19 @@
 }
 
 if test "$_directfb" = yes; then
-  cat > $TMPC << EOF
-#include <directfb_version.h>
-int
-dfb_ver = DIRECTFB_MAJOR_VERSION.DIRECTFB_MINOR_VERSION.DIRECTFB_MICRO_VERSION
-;
-EOF
-  if $_cc -E $TMPC $_inc_extra > "$TMPO"; then
-    _directfb_version=`sed -n 's/^dfb_ver[^0-9]*\(.*\)/\1/p' "$TMPO" | tr -d '()'`
+    _directfb_version=`pkg-config --modversion directfb`
     _dfb_major=`echo $_directfb_version | cut -d . -f 1`
     _dfb_minor=`echo $_directfb_version | cut -d . -f 2`
     _dfb_micro=`echo $_directfb_version | cut -d . -f 3`
     _dfb_version=$(dfb_version $_dfb_major $_dfb_minor $_dfb_micro)
-    if test "$_dfb_version" -ge $(dfb_version 0 9 13); then
+    _def_directfb_version="#define DIRECTFBVERSION $_directfb_version"
+  if pkg-config --atleast-version='0.9.13' directfb ; then
 	_def_directfb_version="#define DIRECTFBVERSION $_dfb_version"
 	_res_comment="$_directfb_version"
     else
         _def_directfb_version='#undef DIRECTFBVERSION'
 	_directfb=no
 	_res_comment="version >=0.9.13 required"
-    fi
-  else
-    _directfb=no
-    _res_comment="failed to get version"
   fi
 fi
 echores "$_directfb"
